### DISCUSSION TITLE
Recommendations for limiting row reads with pagination

### DISCUSSION THREAD

**Discussion by joshiain** [planetscale/discussion#427] [Q&A] [answered]
What's are some recommended ways to implement pagination while being efficient with row reads?

We've so far tried Limit and Offset with a deferred join e.g.

```sql
SELECT * FROM posts
INNER JOIN(select id from posts WHERE id=<user_id> ORDER BY created_at DESC LIMIT 50 OFFSET 10000)
AS lim ON posts.id = lim.id;
```

But as expected this reads all rows that match `WHERE id = <user_id>`.

We've also added a WHERE clause with an index for the created_at column e.g. `WHERE id = <user_id> AND YEAR(created_at) = <year> AND MONTH(created_at) = <month_num>` to at least reduce the number of rows read by the query, but this doesn't reduce it any more than just `WHERE id = <user_id>`. 

If possible we'd like to avoid cursor based pagination to reduce complexity. Is there another type of index or something we can add to improve on row reads? Because if it's going to read all those rows anyway, for a reasonable number of rows, we may as well not bother to limit rows in the call to the database, and instead paginate on the client side to save on row reads.

**Comment by mscoutermarsh** [comment, marked as answer]
The only way I know of would be using cursor pagination.

`LIMIT 50 OFFSET 10000` essentially works by grabbing 10000 + 50 rows and throwing out the first `10000`.

Generally I prefer the usability of LIMIT/OFFSET pagination as well. But it does become a problem with a lot of pages.

I've gone through this decision process a few times on different apps. Here's what I usually look at:
How far back are your users paginating? Offset of 10k would suggest over 200 pages? Is that a common use case?

If it is common, I'd suggest looking more seriously into cursor based just for the performance improvements. If it's rare, then personally I'd leave it as OFFSET/LIMIT. The small number of slow pages can be worth the usability for the 99% of users who don't paginate at all.
